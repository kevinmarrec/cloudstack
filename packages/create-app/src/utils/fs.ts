import fs from 'node:fs/promises'
import path from 'node:path'

import type { PathLike } from 'node:fs'

const ignorePredicate = (filename: string) => ['.git'].includes(filename)

async function empty(dir: string) {
  for (const entry of await fs.readdir(dir)) {
    if (ignorePredicate(entry)) {
      continue
    }

    await fs.rm(path.resolve(dir, entry), { recursive: true })
  }
}

async function emptyCheck(path: PathLike) {
  return fs.readdir(path)
    .then(files => files.every(ignorePredicate))
    .catch(() => true)
}

export async function exists(path: PathLike) {
  return fs.access(path)
    .then(() => true)
    .catch(() => false)
}

export default {
  ...fs,
  empty,
  emptyCheck,
  exists,
}
