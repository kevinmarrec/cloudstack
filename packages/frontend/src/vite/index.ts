/// <reference types="vite-ssg" />

import { readFileSync } from 'node:fs'
import path from 'node:path'

import Vue from '@vitejs/plugin-vue'
import defu from 'defu'
import Unocss from 'unocss/vite'
import AutoImport from 'unplugin-auto-import/vite'
import Components from 'unplugin-vue-components/vite'
import { VueRouterAutoImports } from 'unplugin-vue-router'
import VueRouter from 'unplugin-vue-router/vite'
import { mergeConfig } from 'vite'
import type { PluginOption, UserConfig } from 'vite'
import { VitePWA } from 'vite-plugin-pwa'
import VueDevTools from 'vite-plugin-vue-devtools'
import Layouts from 'vite-plugin-vue-layouts'
import WebfontDownload from 'vite-plugin-webfont-dl'

interface FrameworkOptions {
  root: string
  autoImports?: Parameters<typeof AutoImport>[0]
  components?: Parameters<typeof Components>[0]
  layouts?: Parameters<typeof Layouts>[0]
  pwa?: Parameters<typeof VitePWA>[0]
  unocss?: Parameters<typeof Unocss>[0]
  vue?: Parameters<typeof Vue>[0]
  vueRouter?: Parameters<typeof VueRouter>[0]
  vueDevTools?: Parameters<typeof VueDevTools>[0]
  webfontDownload?: Parameters<typeof WebfontDownload>[0]
}

function applyDefaults(options: FrameworkOptions) {
  options.autoImports = defu(options.autoImports, {
    dts: 'src/types/auto-imports.d.ts',
    imports: [
      'vue',
      VueRouterAutoImports,
      {
        '@unhead/vue': [
          'useHead',
          'useHeadSafe',
          'useSeoMeta',
        ],
      },
      '@vueuse/core',
    ],
    dirs: [
      'src/composables',
      'src/stores',
    ],
  } satisfies FrameworkOptions['autoImports'])

  options.components = defu(options.components, {
    dts: 'src/types/components.d.ts',

  } satisfies FrameworkOptions['components'])

  options.pwa = defu(options.pwa, {
    registerType: 'autoUpdate',
    manifest: {
      name: 'Vite PWA',
      short_name: 'Vite PWA',
      theme_color: '#ffffff',
    },
    pwaAssets: {
      overrideManifestIcons: true,
    },
    devOptions: {
      enabled: true,
      type: 'module',
    },
  } satisfies FrameworkOptions['pwa'])

  options.vueRouter = defu(options.vueRouter, {
    dts: 'src/types/router.d.ts',
  } satisfies FrameworkOptions['vueRouter'])
}

export default function framework(options: FrameworkOptions): PluginOption[] {
  applyDefaults(options)

  return [
    // https://github.com/antfu/unplugin-auto-import
    AutoImport(options.autoImports),

    // https://github.com/antfu/unplugin-vue-components
    Components(options.components),

    // https://github.com/JohnCampionJr/vite-plugin-vue-layouts
    Layouts(options.layouts),

    // https://github.com/unocss/unocss
    Unocss(options.unocss),

    // https://github.com/antfu/vite-plugin-pwa
    VitePWA(options.pwa),

    // https://github.com/vitejs/vite-plugin-vue
    Vue(options.vue),

    // https://github.com/posva/unplugin-vue-router
    VueRouter(options.vueRouter),

    // https://github.com/webfansplz/vite-plugin-vue-devtools
    VueDevTools(options.vueDevTools),

    // https://github.com/feat-agency/vite-plugin-webfont-dl
    WebfontDownload(options.webfontDownload),

    // Additional configuration
    {
      config(config) {
        return mergeConfig<UserConfig, UserConfig>(config, {
          resolve: {
            alias: {
              '~': path.resolve(options.root, 'src'),
            },
          },
          optimizeDeps: {
            include: [
              '@unhead/vue',
              '@vueuse/core',
              'workbox-window',
            ],
          },
          ssgOptions: {
            script: 'async',
            formatting: 'minify',
          },
        })
      },
      transformIndexHtml(html) {
        return {
          html,
          tags: [
            {
              tag: 'script',
              children: readFileSync(path.join(import.meta.url.replace('file:', ''), '../darkMode.script.js'), 'utf-8'),
              injectTo: 'head',
            },
          ],
        }
      },
    },
  ]
}
