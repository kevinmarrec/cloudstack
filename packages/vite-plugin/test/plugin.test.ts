import { mkdir, mkdtemp, rm, writeFile } from 'node:fs/promises'
import os from 'node:os'
import path from 'node:path'
import process from 'node:process'

import { createServer, resolveConfig } from 'vite'
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'

import CloudstackVitePlugin, { type CloudstackPluginOptions } from '../src'

async function createTempDir() {
  const osTmpDir = os.tmpdir()
  const tmpDir = path.resolve(osTmpDir, 'vite-plugin-test')
  return await mkdtemp(tmpDir)
}

const mocks = vi.hoisted(() => ({
  isPackageExists: vi.fn(),
}))

vi.mock('local-pkg', () => ({
  isPackageExists: mocks.isPackageExists,
}))

describe('plugin', () => {
  let tmpDir: string

  beforeEach(async () => {
    tmpDir = await createTempDir()
  })

  afterEach(async () => {
    await rm(tmpDir, { recursive: true, force: true })
  })

  it('with default options', async () => {
    const baseConfig = await resolveConfig({ root: tmpDir }, 'serve')

    vi.spyOn(process, 'cwd').mockReturnValue(tmpDir)

    const resolvedConfig = await resolveConfig({
      root: tmpDir,
      plugins: [CloudstackVitePlugin()],
    }, 'serve')

    const addedPlugins = resolvedConfig.plugins.filter(plugin => !baseConfig.plugins.some(basePlugin => basePlugin.name === plugin.name))
    const addedPluginNames = addedPlugins.map(plugin => plugin.name)

    expect(addedPluginNames).toMatchSnapshot()
  })

  it('with all options', async () => {
    const baseConfig = await resolveConfig({ root: tmpDir }, 'serve')

    vi.spyOn(process, 'cwd').mockReturnValue(tmpDir)

    mocks.isPackageExists.mockImplementation((pkg: string) => pkg === 'vue-router' || pkg === 'unocss')

    await mkdir(path.resolve(tmpDir, 'src/components'), { recursive: true })
    await writeFile(path.resolve(tmpDir, 'src/components/HelloWorld.vue'), `<template> <h1>Hello World</h1> </template>`)

    await mkdir(path.resolve(tmpDir, 'src/layouts'), { recursive: true })
    await writeFile(path.resolve(tmpDir, 'src/layouts/default.vue'), `<template> <slot /> </template>`)

    await mkdir(path.resolve(tmpDir, 'src/pages'), { recursive: true })
    await writeFile(path.resolve(tmpDir, 'src/pages/Index.vue'), `<template> <h1>Index</h1> </template>`)

    await writeFile(path.resolve(tmpDir, 'uno.config.ts'), `export default {}`)

    const resolvedConfig = await resolveConfig({
      root: tmpDir,
      plugins: [CloudstackVitePlugin({
        pwa: {
          pwaAssets: {
            disabled: true,
          },
        },
      })],
    }, 'serve')

    const addedPlugins = resolvedConfig.plugins.filter(plugin => !baseConfig.plugins.some(basePlugin => basePlugin.name === plugin.name))
    const addedPluginNames = addedPlugins.map(plugin => plugin.name)

    expect(addedPluginNames).toMatchSnapshot()
  })

  it('should optimize dependency "workbox-window" with pwa option', async () => {
    const resolvedConfig = await resolveConfig({
      root: tmpDir,
      plugins: [
        CloudstackVitePlugin({
          pwa: {
            pwaAssets: {
              disabled: true,
            },
          },
        }),
      ],
    }, 'build')

    expect(resolvedConfig.optimizeDeps.include).toEqual(
      expect.arrayContaining(['workbox-window']),
    )
  })

  it('should inject dark mode script in index.html', async () => {
    const server = await createServer({
      root: tmpDir,
      plugins: [
        CloudstackVitePlugin(),
      ],
    })

    const html = await server.transformIndexHtml('index.html', '<html></html>')

    expect(html).toMatchSnapshot()
  })
})
