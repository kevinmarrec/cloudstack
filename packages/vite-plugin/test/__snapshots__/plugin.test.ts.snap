// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugin > should inject dark mode script in index.html 1`] = `
"<html>
<script>(function () {
  const prefersDark = window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches
  const setting = localStorage.getItem('vueuse-color-scheme') || 'auto'
  if (setting === 'dark' || (prefersDark && setting !== 'light'))
    document.documentElement.classList.toggle('dark', true)
})()</script>

<script type="module" src="/@id/virtual:vue-devtools-path:overlay.js"></script>
<script type="module" src="/@id/virtual:vue-inspector-path:load.js"></script>

<script type="module" src="/@vite/client"></script>
</html>"
`;

exports[`plugin > with all integrations 1`] = `
[
  "vite-plugin-pwa",
  "unocss:transformers:pre",
  "unocss:global:content",
  "unocss:global",
  "unplugin-vue-router",
  "vite-plugin-inspect",
  "vite-plugin-vue-devtools",
  "vite:cloudstack:global",
  "vite:cloudstack:config",
  "vite:cloudstack:dark-mode",
  "vite-plugin-vue-layouts",
  "vite-plugin-pwa:dev-sw",
  "unocss:config",
  "unocss:transformers:undefined",
  "unocss:devtools",
  "unocss:api",
  "unocss:inspector",
  "vite:vue",
  "unplugin-auto-import",
  "unplugin-vue-components",
  "vite-plugin-pwa:info",
  "vite-plugin-pwa:pwa-assets",
  "unocss:transformers:post",
  "unocss:global:post",
]
`;

exports[`plugin > with options: { autoImports: false } 1`] = `
[
  "vite-plugin-inspect",
  "vite-plugin-vue-devtools",
  "vite:cloudstack:global",
  "vite:cloudstack:config",
  "vite:cloudstack:dark-mode",
  "vite:vue",
]
`;

exports[`plugin > with options: { devtools: false } 1`] = `
[
  "vite:cloudstack:global",
  "vite:cloudstack:config",
  "vite:cloudstack:dark-mode",
  "vite:vue",
  "unplugin-auto-import",
]
`;

exports[`plugin > with options: {} 1`] = `
[
  "vite-plugin-inspect",
  "vite-plugin-vue-devtools",
  "vite:cloudstack:global",
  "vite:cloudstack:config",
  "vite:cloudstack:dark-mode",
  "vite:vue",
  "unplugin-auto-import",
]
`;

exports[`virtual module > should generate virtual module content, with options: { router: {} } 1`] = `
"import { ViteSSG } from 'vite-ssg'
import { routes } from 'vue-router/auto-routes'
import 'the-new-css-reset'
export const Power = (App, fn) => ViteSSG(App, { base: import.meta.env.BASE_URL, routes }, fn)"
`;

exports[`virtual module > should generate virtual module content, with options: { router: {}, layouts: {} } 1`] = `
"import { ViteSSG } from 'vite-ssg'
import { routes } from 'vue-router/auto-routes'
import { setupLayouts } from 'virtual:generated-layouts'
import 'the-new-css-reset'
export const Power = (App, fn) => ViteSSG(App, { base: import.meta.env.BASE_URL, routes: setupLayouts(routes) }, fn)"
`;

exports[`virtual module > should generate virtual module content, with options: { unocss: {} } 1`] = `
"import { ViteSSG } from 'vite-ssg/single-page'
import 'the-new-css-reset'
import 'uno.css'
export const Power = (App, fn) => ViteSSG(App, fn)"
`;

exports[`virtual module > should generate virtual module content, with options: {} 1`] = `
"import { ViteSSG } from 'vite-ssg/single-page'
import 'the-new-css-reset'
export const Power = (App, fn) => ViteSSG(App, fn)"
`;
